api/src/main/java/bio/terra/tanagra/indexing/Indexer.java

```diff
@@ -0,0 +1,106 @@
+package bio.terra.tanagra.indexing;
+
+import bio.terra.tanagra.serialization.UFEntity;
+import bio.terra.tanagra.serialization.UFEntityGroup;
+import bio.terra.tanagra.serialization.UFUnderlay;
+import bio.terra.tanagra.underlay.Entity;
+import bio.terra.tanagra.underlay.Underlay;
+import bio.terra.tanagra.utils.FileUtils;
+import bio.terra.tanagra.utils.JacksonMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class Indexer {
+  private static final Logger LOGGER = LoggerFactory.getLogger(Indexer.class);
+  public static final String OUTPUT_UNDERLAY_FILE_EXTENSION = ".json";
+  public static final Function<String, InputStream> READ_RESOURCE_FILE_FUNCTION =
+      filePath -> FileUtils.getResourceFileStream(filePath);
+  public static final Function<String, InputStream> READ_FILE_FUNCTION =
+      filePath -> FileUtils.getFileStream(filePath);
+
+  private final String underlayPath;
+  private final Function<String, InputStream> getFileInputStreamFunction;
+
+  private List<WorkflowCommand> indexingCmds;
+  private UFUnderlay expandedUnderlay;
+  private List<UFEntity> expandedEntities;
+  private List<UFEntityGroup> expandedEntityGroups;
+
+  private Indexer(String underlayPath, Function<String, InputStream> getFileInputStreamFunction) {
+    this.underlayPath = underlayPath;
+    this.getFileInputStreamFunction = getFileInputStreamFunction;
+  }
+
+  public static Indexer fromResourceFile(String underlayResourceFilePath) {
+    return new Indexer(underlayResourceFilePath, READ_RESOURCE_FILE_FUNCTION);
+  }
+
+  public static Indexer fromFile(String underlayFilePath) {
+    return new Indexer(underlayFilePath, READ_FILE_FUNCTION);
+  }
+
+  public void indexUnderlay() throws IOException {
+    // deserialize the POJOs to the internal objects and expand all defaults
```

Comment:
OK, I'm happy to use this convention going forward for this repo. I'll reformat comments as I come across them, instead of going through all existing comments.

Reply:
Sounds good.